deeds-dapp-common/src/main/java/io/meeds/deeds/service/AuthorizationCodeService.java

```diff
@@ -97,13 +106,32 @@ public void init() {
    *                                  reached or maximum code verification has
    *                                  been reached
    */
-  public void generateCode(String key, String email, Object data) throws IllegalAccessException {
+  public void generateCode(String key, String email, Object data, String clientIp) throws IllegalAccessException {
+    long currentTimestamp = System.currentTimeMillis() / 1000;
+    int count = 0;
     if (!hasValidDateCode(key)) {
       authorizationCodes.put(key, newAuthorizationCode());
     }
+    Iterator<Map.Entry<String, Long>> iterator = codeGenerationStamp.entrySet().iterator();
+    iterator.forEachRemaining(entry -> {
+            boolean olderThanOneMinute = currentTimestamp - entry.getValue() > 60;
+            if (olderThanOneMinute) {
+                codeGenerationCount.remove(entry.getKey());
+                iterator.remove();
+            }
+        });
+    if (codeGenerationCount.containsKey(clientIp)) {
+      count = codeGenerationCount.get(clientIp);
+      if (count >= generationCodeLimit) {
+        throw new IllegalAccessException("Code generation limit exceeded");
+      }
+    } else {
+      codeGenerationStamp.put(clientIp,currentTimestamp);
+    }
```

Comment:
Externalize this in a dedicated method `checkClientIpBlackListed` + **throw the exception** with the ip address when blacklisted.

Reply:
done!